lib/ddtrace/contrib/rails/core_extensions.rb

```diff
@@ -157,51 +157,34 @@ def patch_action_controller
 
     def patch_process_action
       do_once(:patch_process_action) do
+        ActionControllerPatch.instrument(::ActionController::Metal)
+      end
+    end
+
+    # Compat module for Ruby versions not supporting #prepend
+    module CompatActionControllerPatch
+      def process_action(*args)
+        process_action_without_datadog(*args)
+      end
+    end
+
+    # ActionController patch
+    module ActionControllerPatch
+      def self.instrument(klass)
```

Comment:
The strategy here is good, but it isn't quite as idiomatic as it could be. I might suggest restructuring this a bit to take advantage of the Ruby `included` method to drive this. Something like:

```ruby
module ActionControllerPatch
  def self.included(base)
    if Gem::Version.new(RUBY_VERSION) >= Gem::Version.new('2.0.0')
      base.send(:prepend, ProcessActionPatch)
    else
      base.class_eval do
        alias_method :process_action_without_datadog, :process_action

        include ProcessActionPatch
      end
    end
  end

  module ProcessActionCompatibilityPatch
    def process_action(*args)
      process_action_without_datadog(*args)
    end
  end

  module ProcessActionPatch
    # compat for Ruby versions not supporting #prepend
    include ProcessActionCompatibilityPatch unless Gem::Version.new(RUBY_VERSION) >= Gem::Version.new('2.0.0')

    def process_action(*args)
      # ...
    end
  end
end
```

This would enable you to simply do `::ActionController::Metal.send(:include, ActionControllerPatch)` to apply the patch.

Some benefits of doing so:

 - Looks a bit more idiomatic
 - You effectively prevent patching the controller twice, by using `include` instead of a generic module function.
 - You can compartmentalize all of the modules needed to patch `ActionController` within the `ActionControllerPatch` module.

Reply:
❤️